### DISCUSSION TITLE
How to use DI to use a DbContext in a wrapper class, and inject said wrapper class instead of the DbContext

### DISCUSSION THREAD

**Discussion by IP14Y3RI** [dotnet/aspnetcore#24110] [General]
Hi!
I am working on a API and want to utilise the Dependency Injection functionality that is built into ASP.NET Core. However, I am struggling to solve a seemingly simple problem. I am using Entity Framework Core to connect with my MySQL database, and unfortunately, I couldn't exactly find a possible answer to my question.

So I have configured a DbContext.
```
public class MyContect : DbContext
    {
        public  MyContect (DbContextOptions<MyContect> options) : base(options)
        {
        }
    }
```

And then in the `Startup.cs` file, in the `ConfigureServices` method, I have
```
services.AddDbContext<MyContext>(options => options.UseMySQL(
                   Configuration.GetConnectionString("DefaultConnection")));
```

Now in my Controller, I have

```
public class MyController : ControllerBase{
        private readonly DAO _dao;
        private readonly MyContext _context;

        public AuthController(AuthContext context, )
        {
            this._context = context;
        }
}
```

Now the instance of MyContext is injected into my controller just fine and it all works.

For testing purposes, I have introduced a new class `DAO`, which uses a `MyContext` instance. By doing so, I want that `MyController` can get injected an instance of `DAO` instead of getting injected `MyContext` directly , and I am aiming to inject an instance of `MyContext` into `DAO`. In other words; The instance of `DAO` should use/receive the instanced of `MyContext`, and `MyController` should receive the instance of `DAO`.

Then, the constructor of `MyController` should look like something like this:
```
public class MyController : ControllerBase{
        private readonly DAO _dao;
       // private readonly MyContext _context; // No longer needed.

        public AuthController(DAO Dao )
        {
            this._dao = dao;
        }
}
```

And my `DAO` class would look like something like this
```
public class DataAccessObject {
        private readonly AuthContext _context;

        public DataAccessObject(AuthContext context)
        {
            _context = context;
        }

. . .
}
```

How can I achieve this? What methods do I need to use in `Startup.ConfigureServices` to do this?

**Comment by davidfowl** [comment]
You have to tell the container about your DataAccessObject:

```C#
services.AddTransient<DataAccessObject>();
```

Read more about dependency injection here https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1
